File: androlua/src/main/java/cn/vove7/androlua/luabridge/LuaAsyncTask.java
package cn.vove7.androlua.luabridge;

import android.os.AsyncTask;

import cn.vove7.androlua.luautils.LuaGcable;
import com.luajava.JavaFunction;
import com.luajava.LuaException;
import com.luajava.LuaObject;
import com.luajava.LuaState;

import cn.vove7.androlua.LuaApp;
import cn.vove7.androlua.luautils.LuaManagerI;

import static cn.vove7.androlua.luabridge.LuaUtil.errorReason;

public class LuaAsyncTask extends AsyncTask implements LuaGcable {

    private Object[] loadeds;
    private LuaState L;

    private LuaManagerI luaManager;
    private byte[] mBuffer;
    private long mDelay = 0;
    private LuaObject mCallback;
    private LuaObject mUpdate;

    public LuaAsyncTask(LuaManagerI luaManager, long delay, LuaObject callback) throws LuaException {
        LuaApp.getInstance().regGc(this);
        this.luaManager = luaManager;
        mDelay = delay;
        mCallback = callback;
    }

    public LuaAsyncTask(LuaManagerI luaManager, String src, LuaObject callback) throws LuaException {
        luaManager.regGc(this);
        this.luaManager = luaManager;
        mBuffer = src.getBytes();
        mCallback = callback;
    }

    public LuaAsyncTask(LuaManagerI luaManager, LuaObject func, LuaObject callback) throws LuaException {
        luaManager.regGc(this);
        this.luaManager = luaManager;
        mBuffer = func.dump();
        mCallback = callback;
        LuaState l = func.getLuaState();
        LuaObject g = l.getLuaObject("luajava");
        LuaObject loaded = g.getField("imported");
        if (!loaded.isNil()) {
            loadeds = loaded.asArray();
        }
    }


    public LuaAsyncTask(LuaManagerI luaManager, LuaObject func, LuaObject update, LuaObject callback) throws LuaException {
        this.luaManager = luaManager;
        luaManager.regGc(this);
        mBuffer = func.dump();
        mUpdate = update;
        mCallback = callback;
    }

    @Override
    public void gc() {
        if (getStatus() == Status.RUNNING)
            cancel(true);
    }

    public void execute() throws IllegalArgumentException, ArrayIndexOutOfBoundsException, LuaException {
        super.execute();
    }

    public void update(Object msg) {
        publishProgress(msg);
    }

    public void update(String msg) {
        publishProgress(msg);
    }

    public void update(int msg) {
        publishProgress(msg);
    }

    @Override
    protected Object doInBackground(Object[] args) {
        if (mDelay != 0) {
            try {
                Thread.sleep(mDelay);
            } catch (InterruptedException e) {
            }
            return args;
        }
        L = LuaApp.getInstance().getLuaHelper().L;

        try {
            JavaFunction update = new JavaFunction(L) {
                @Override
                public int execute() throws LuaException {
                    update(L.toJavaObject(2));
                    return 0;
                }
            };
            update.register("update");
        } catch (LuaException e) {
            e.printStackTrace();
            luaManager.log("AsyncTask" + e.getMessage());
        }

        if (loadeds != null) {
            LuaObject require = L.getLuaObject("require");
            try {
                require.call("import");
                LuaObject _import = L.getLuaObject("import");
                for (Object s : loadeds)
                    _import.call(s.toString());
            } catch (LuaException e) {
                e.printStackTrace();
            }
        }

        try {
            L.setTop(0);
            int ok = L.LloadBuffer(mBuffer, "LuaAsyncTask");

            if (ok == 0) {
                L.getGlobal("debug");
                L.getField(-1, "traceback");
                L.remove(-2);
                L.insert(-2);
                int l = args.length;
                for (Object arg : args) {
                    L.pushObjectValue(arg);
                }
                ok = L.pcall(l, LuaState.LUA_MULTRET, -2 - l);
                if (ok == 0) {
                    int n = L.getTop() - 1;
                    Object[] ret = new Object[n];
                    for (int i = 0; i < n; i++)
                        ret[i] = L.toJavaObject(i + 2);
                    return ret;
                }
            }
            throw new LuaException(errorReason(ok) + ": " + L.toString(-1));
        } catch (LuaException e) {
            e.printStackTrace();
            luaManager.handleMessage(LuaManagerI.E,"doInBackground" + e);
        }
        return null;
    }

    @Override
    protected void onPostExecute(Object result) {
        if (isCancelled())
            return;
        try {
            if (mCallback != null)
                mCallback.call((Object[]) result);
        } catch (LuaException e) {
            luaManager.handleMessage(LuaManagerI.E,"onPostExecute" + e.toString());
        }
        super.onPostExecute(result);
        if (L != null)
            L.gc(LuaState.LUA_GCCOLLECT, 1);
        System.gc();
        //L.close();
    }

    @Override
    protected void onProgressUpdate(Object[] values) {
        // TODO: Implement this method
        try {
            if (mUpdate != null)
                mUpdate.call(values);
        } catch (LuaException e) {
            luaManager.handleMessage(LuaManagerI.E, "onProgressUpdate" + e.getMessage());
        }
        super.onProgressUpdate(values);
    }

}

File: androlua/src/main/java/cn/vove7/androlua/luabridge/LuaAsyncTask.kt
package cn.vove7.androlua.luabridge

import android.os.AsyncTask
import cn.vove7.androlua.LuaApp
import cn.vove7.androlua.LuaHelper
import cn.vove7.androlua.luautils.LuaGcable
import cn.vove7.androlua.luautils.LuaManagerI
import cn.vove7.androlua.luautils.LuaRunnableI
import cn.vove7.common.executor.OnPrint
import cn.vove7.vtp.log.Vog
import com.luajava.JavaFunction
import com.luajava.LuaException
import com.luajava.LuaObject
import com.luajava.LuaState


class LuaAsyncTask : AsyncTask<Any, Any, Any>, LuaGcable, LuaRunnableI, Comparable<LuaAsyncTask> {
    override fun compareTo(other: LuaAsyncTask): Int = hashCode() - other.hashCode()
    private var L: LuaState

    private var luaManager: LuaManagerI
    private var mBuffer: ByteArray? = null
    private var mDelay: Long = 0
    private var mCallback: LuaObject? = null
    private var mUpdate: LuaObject? = null

    private var funHelper: LuaFunHelper
    private var luaHelper: LuaHelper

    @Throws(LuaException::class)
    constructor(luaManager: LuaManagerI, delay: Long, callback: LuaObject) : this(luaManager, callback) {
        mDelay = delay
    }

    @Throws(LuaException::class)
    constructor(luaManager: LuaManagerI, src: String, callback: LuaObject) : this(luaManager, callback) {
        mBuffer = src.toByteArray()
    }

    @Throws(LuaException::class)
    constructor(luaManager: LuaManagerI, func: LuaObject, callback: LuaObject) : this(luaManager, callback) {
        mBuffer = func.dump()
    }


    @Throws(LuaException::class)
    constructor(luaManager: LuaManagerI, func: LuaObject, update: LuaObject,
                callback: LuaObject) : this(luaManager, callback) {
        mBuffer = func.dump()
        mUpdate = update
    }

    constructor(luaManager: LuaManagerI, callback: LuaObject) {
        this.luaManager = luaManager
        luaManager.regGc(this)
        mCallback = callback

        luaHelper = LuaHelper(LuaApp.instance)
        L = luaHelper.L
        funHelper = LuaFunHelper(luaHelper, L)

        funHelper.copyRuntimeFrom(luaManager.luaState)
    }

    override fun quit() {
        quit(false)
    }

    override fun quit(self: Boolean) {
        Vog.d(this, "quit $this $self")
        luaManager.removeGc(this)
        L.gc(LuaState.LUA_GCCOLLECT, 1)
        System.gc()
//        L.close()
        if (status == AsyncTask.Status.RUNNING) {
            Vog.d(this, "cancel")
            cancel(true)
        }
    }

    override fun gc() {
        quit(false)
    }

    @Throws(IllegalArgumentException::class, ArrayIndexOutOfBoundsException::class, LuaException::class)
    fun exec() {
        super.execute()
    }

    fun cancel() {
        quit()
        cancel(true)
    }

    fun exec(args: Array<Any>) {
        Vog.d(this, "exec $args")
        super.execute(*args)
    }


    fun update(msg: Any?) {
        publishProgress(msg)
    }

    fun update(msg: String) {
        publishProgress(msg)
    }

    fun update(msg: Int) {
        publishProgress(msg)
    }

    override fun doInBackground(args: Array<Any>): Any? {
        if (mDelay != 0L) {
            try {
                Thread.sleep(mDelay)
            } catch (e: InterruptedException) {
                e.printStackTrace()
                luaManager.handleError(e)
                quit()
                return null
            }
            return args
        }

        try {
            val update = object : JavaFunction(L) {
                @Throws(LuaException::class)
                override fun execute(): Int {
                    update(L.toJavaObject(2))
                    return 0
                }
            }
            update.register("update")
        } catch (e: LuaException) {
            e.printStackTrace()
            luaManager.handleError(e)
        }
        L.top = 0
        var ok = L.LloadBuffer(mBuffer, "LuaAsyncTask")

        if (ok == 0) {
            L.getGlobal("debug")
            L.getField(-1, "traceback")
            L.remove(-2)
            L.insert(-2)
            val l = args.size
            for (arg in args) {
                L.pushObjectValue(arg)
            }
            ok = L.pcall(l, LuaState.LUA_MULTRET, -2 - l)
            if (ok == 0) {
                val n = L.top - 1
                val ret = arrayOfNulls<Any>(n)
                for (i in 0 until n)
                    ret[i] = L.toJavaObject(i + 2)
                return ret
            } else luaHelper.checkErr(ok)
        }
        return null
    }

    override fun onPostExecute(result: Any) {
        if (isCancelled)
            return
        try {
            if (mCallback != null)
                mCallback!!.call(*result as Array<Any>)
        } catch (e: LuaException) {
            luaManager.handleMessage(OnPrint.ERROR, "onPostExecute" + e.toString())
        }
        super.onPostExecute(result)
        quit(true)
    }

    protected override fun onProgressUpdate(values: Array<Any>) {
        try {
            mUpdate?.call(*values)
        } catch (e: LuaException) {
            e.printStackTrace()
            luaManager.handleMessage(OnPrint.ERROR, "onProgressUpdate" + e.message)
        }
//        super.onProgressUpdate(*values)
    }

}

